File: lib/indicators.ts
// Technical indicators precomputed over a bar series.

import type { Bar } from "./data";

export interface Indicators {
  rsi: number[]; // length == bars.length, initial values default to 50
  logRet: number[]; // length == bars.length, first value 0
}

export function computeIndicators(bars: Bar[], rsiPeriod = 14): Indicators {
  const n = bars.length;
  const logRet = new Array<number>(n).fill(0);
  for (let i = 1; i < n; i++) {
    logRet[i] = Math.log(bars[i].close / bars[i - 1].close);
  }

  const rsi = new Array<number>(n).fill(50);
  let avgGain = 0;
  let avgLoss = 0;
  const alpha = 1 / rsiPeriod;
  for (let i = 1; i < n; i++) {
    const delta = bars[i].close - bars[i - 1].close;
    const gain = Math.max(delta, 0);
    const loss = Math.max(-delta, 0);
    avgGain = avgGain + alpha * (gain - avgGain);
    avgLoss = avgLoss + alpha * (loss - avgLoss);
    if (avgLoss === 0) {
      rsi[i] = 100;
    } else {
      const rs = avgGain / avgLoss;
      rsi[i] = 100 - 100 / (1 + rs);
    }
  }
  return { rsi, logRet };
}

File: tradebot/indicators.py
"""Lightweight technical indicators used for state features."""
from __future__ import annotations

import numpy as np
import pandas as pd


def rsi(close: pd.Series, period: int = 14) -> pd.Series:
    delta = close.diff()
    gain = delta.clip(lower=0.0)
    loss = -delta.clip(upper=0.0)
    avg_gain = gain.ewm(alpha=1 / period, adjust=False).mean()
    avg_loss = loss.ewm(alpha=1 / period, adjust=False).mean()
    rs = avg_gain / avg_loss.replace(0.0, np.nan)
    out = 100.0 - (100.0 / (1.0 + rs))
    return out.fillna(50.0)


def log_returns(close: pd.Series, periods: int = 1) -> pd.Series:
    return np.log(close).diff(periods).fillna(0.0)


def sma(close: pd.Series, window: int) -> pd.Series:
    return close.rolling(window=window, min_periods=1).mean()
